lib/active_merchant/billing/gateways/payflow.rb

```diff
@@ -278,6 +282,24 @@ def add_credit_card(xml, credit_card, options = {})
         end
       end
 
+      def add_stored_credential(xml, stored_credential)
+        return unless stored_credential
+
+        xml.tag! 'CardOnFile', recurring_code(stored_credential)
+        xml.tag! 'TxnId', stored_credential[:network_transaction_id] if stored_credential[:network_transaction_id]
+      end
+
+      def recurring_code(stored_credential)
```

Comment:
Thanks for your suggestions. I have updated all these points 👍

Reply:
@ajawadmirza I was thinking about this some more last night. If we recognize that `initial_transaction = true` means that the transaction is customer-initiated, we might be able to simply treat the `CITU` case as a special case and then use a case statement to handle all the other options. A refactored version of that approach would look like:

```
def card_on_file_type(stored_credential)
  if stored_credential[:initial_transaction]
    return 'CITU' if customer_initiated_unscheduled?(stored_credential)

    initial_transaction_type(stored_credential[:reason_type])
  else
    recurring_transaction_type(stored_credential[:reason_type])
  end
end

def customer_initiated_unscheduled?(stored_credential)
  stored_credential[:initiator] == 'cardholder' && stored_credential[:reason_type] == 'unscheduled'
end

def initial_transaction_type(reason_type)
  case reason_type
  when 'unscheduled'
    'CITI'
  when 'recurring', 'installment'
    'CITR'
  end
end

def recurring_transaction_type(reason_type)
  case reason_type
  when 'unscheduled'
    'MITU'
  when 'recurring', 'installment'
    'MITR'
  end
end
```

What do you think? The advantage here is that it may be easier for somebody unfamiliar with the code to look at this and see the values that end up being added to the request.

Let me know which approach you prefer and I will re-review.

Reply:
Thanks @dsmcclain , this approach looks good to me as the documentation mentions that the subsequent transactions will be performed by merchant only. I am just bit unclear about **CITI** and **CITU** transaction types in this approach, it looks like we will only be able to send **CITU** and CITI will never be sent on the request as they both have same conditions?

Reply:
Oh, good point @ajawadmirza.

I think perhaps leaving it as we have it will be sufficient. Sorry to delay the review.